### ISSUE TITLE
Extending the byte_fallback option

### ISSUE THREAD

**Issue by chris-ha458** [google/sentencepiece#852]
As currently implemented the byte_fallback option adds 256 tokens in the form of 
<0x00> <0x01> up to <0xFF> after the bos,eos,pad pieces and after the user_defined_symbols

If we assume that the inputs are all in the form of UTF-8, there are multiple constraints to the possible bytes.

For instance,
1. A continuation byte is invalid unless following a leading byte. 
2. Leading bytes should be followed by the correct number of continuation bytes. (1~3)

Obviously sentencepiece is not a unicode decoder and shouldn't try to be.
However, there are multiple bytes can be easily identified as erroneous or spurious.

For instance if it is possible to have tokens that would look like

▁(\u2581)<0x80>  (continuation byte without a leading byte)
<0xD4>▁ (leading byte without a continuation byte)

it would be obvious that it is problematic since it is invalid according to unicode rules.
Identifying such tokens might help in a practical level for either the model itself or the decoder for it to identify
such edge cases (etc adding them to the bad_words_ids or assigining them a high negative loss)

Also consider the case such as this

<0xB3>▁ (a continuation byte that not followed by a continuation byte so also terminating code point)

In this case, the token itself maybe invalid or not. To completely tell we would need to implement unicode decoding which is a big task.
It is possible, however, to distinguish itself from a <0xB3> that finishes a codepoint.

In other words, with separate tokens for <0xB3> and <0xB3>▁ , the model would be able to discern at a byte level
When unicode byte sequences that happen to finish with <0xB3>▁ or when it merely appears amongst a multibyte sequence

곳  = 	11101010 **10110011** **10110011**  <0xEA><0xB3><0xB3>▁ 
곰 =         11101010 **10110011** 10110000   <0xEA><0xB3><0xB0>▁ 
갳 = 	11101010 10110000 **10110011** <0xEA><0xB0><0xB3>▁ 
걳 = 	11101010 10110001 **10110011** <0xEA><0xB1><0xB3>▁ 

For the first instance two consequent <0xB3> bytes would be read and treated by the model differently.

In practical terms, the model will have more capacity to embed multilingual unicode characters.
Instead of sharing <0xB3> to both encode semantic information of  곳, 곳 (since it appears twice in 곳)
and 곰 <0xB3> and <0xB3>▁ might "share" the load.
In a purely hypothetical example. <0xB3> can know encode 고 <0xB3>▁ can encode ㅅ to make 곳
while <0xB0>▁ can encode ㅁ to make <0xB3> + <0XB0>▁ = 곰

I know this is a very strained example which may or may not contribute to korean or other multilingual performance
But it would be very awesome if we atleast had an option.
If we consider 2 prefix ▁ options(present, absent)<byte_fallback tokens> 2 suffix ▁ options (present,absent)
Then the total "token budget cost" would be a mere 1024, which is a small price to pay in 100k+ budgets of multilingual models.
Of course with an indepth knowledge of UTF-8 this cost can be further cut 
( no option necessary for first 128 ascii codepoints so that cuts the cost in half, and further reductions possible)

Would it be possible to implement this in the code?
If not, would setting --user_defined_symbols or required_chars to "▁<0xB3>" etc. achieve the effect I am looking for?

**Comment by taku910** [comment]
We can understand that the output may contain invalid utf8 characters, but our observation in real applications says that such cases rarely occur as long as using sufficient training data. Utf8 byte constraints are so trivial/simple for neural networks.  The model can train the UTF8 constraint easily.

We would recommend using more training data or let the training include more byte-fallbacked sequences so the model can capture UT8 constraints.

If we still have the problem, you can emulate the extended utf8 byte sequence with user_defined_symbols.
